libnavigation-core/src/main/java/com/mapbox/navigation/core/arrival/ArrivalProgressObserver.kt

```diff
@@ -34,23 +40,40 @@ internal class ArrivalProgressObserver(
         arrivalObservers.clear()
     }
 
-    fun navigateNextRouteLeg(): Boolean {
+    fun navigateNextRouteLeg(callback: LegIndexUpdatedCallback?) {
         val routeProgress = tripSession.getRouteProgress()
         val numberOfLegs = routeProgress?.route?.legs()?.size
-            ?: return false
+
+        if (numberOfLegs == null) {
+            callback?.onLegIndexUpdatedCallback(false)
+            return
+        }
         val legProgress = routeProgress.currentLegProgress
         val legIndex = legProgress?.legIndex
-            ?: return false
-        val nextLegIndex = legIndex + 1
-        val nextLegStarted = if (nextLegIndex < numberOfLegs) {
-            tripSession.updateLegIndex(nextLegIndex)
-        } else {
-            false
+        if (legIndex == null) {
+            callback?.onLegIndexUpdatedCallback(false)
+            return
         }
-        if (nextLegStarted) {
-            arrivalObservers.forEach { it.onNextRouteLegStart(legProgress) }
+
+        mainJobControl.scope.launch {
+            val route = directionsSession.routes.firstOrNull()
+            val nextLegIndex = legIndex + 1
+            val nextLegStarted = if (nextLegIndex < numberOfLegs) {
+                tripSession.updateLegIndex(nextLegIndex)
+            } else {
+                false
+            }
+
+            val currentRoute = directionsSession.routes.firstOrNull()
+            if (route == currentRoute) {
```

Comment:
Equals comparison also includes annotations, which shouldn't count here because that could be updated via refresh controller.

Thinking more about it, the callback should probably be called with an error only if native `Navigator#setRoutes` was called while this update was running.